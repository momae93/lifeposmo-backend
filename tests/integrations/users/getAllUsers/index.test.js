const axios = require('axios');

const { buildMockDatabase } = require('../../../helpers/');
const { buildGraphqlServer } = require('../../../../src/bin/servers/index');
const { buildServices: buildUserServices } = require('../../../../src/modules/users');
const { buildServices: buildPostServices } = require('../../../../src/modules/posts');
const { URL_GRAPHQL_API_TEST_INTEGRATION } = require('../../constants');
const buildMockUserRepository = require('../../../../src/modules/users/__mocks__/repository');
const buildMockPostRepository = require('../../../../src/modules/posts/__mocks__/repository');

describe('[INTEGRATION][USERS] : Get all user', () => {
  let server = null;
  let mockDatabase = null;

  beforeAll(async () => {
    // LOCAL MOCKS
    const mockDbServices = {};
    const mockUserRepository = buildMockUserRepository({
      getAllUsers: () => (mockDatabase.users),
    });
    const mockPostRepository = buildMockPostRepository({
      getPostsByIdAuthor: (idAuthor) => (mockDatabase
        .posts
        .filter((posts) => posts.idAuthor === idAuthor)),
    });
    const mockBuildUserRepository = () => (mockUserRepository);
    const mockBuildPostRepository = () => (mockPostRepository);
    const userServices = buildUserServices(mockDbServices, mockBuildUserRepository);
    const postServices = buildPostServices(mockDbServices, mockBuildPostRepository);
    const services = {
      userServices,
      postServices,
    };

    // SETUP SERVER
    server = buildGraphqlServer(services);

    // STARTING SERVER
    await server.start();
  });

  beforeEach(() => {
    mockDatabase = buildMockDatabase();
  });

  afterAll(async () => {
    // STOPPING SERVER SERVER
    await server.stop();
  });

  it('should success returns all fields when calling * users * with all its fields', async () => {
    // QUERY
    const GET_ALL_USERS_REQUEST = `
      query {
        users {
          id
          firstname
          lastname
          username
          password
          description
          birthdate
          isMale
        }
      }
    `;

    const { data: users } = await axios
      .post(URL_GRAPHQL_API_TEST_INTEGRATION, { query: GET_ALL_USERS_REQUEST });

    expect(users).toMatchSnapshot();
  });

  it('should success returns some fields when calling * users * with some of its fields', async () => {
    // QUERY
    const GET_ALL_USERS_REQUEST = `
      query {
        users {
          firstname
          lastname
          description
        }
      }
    `;

    const { data: users } = await axios
      .post(URL_GRAPHQL_API_TEST_INTEGRATION, { query: GET_ALL_USERS_REQUEST });

    expect(users).toMatchSnapshot();
  });

  it('should success returns all fields when calling * users * with their postwritten', async () => {
    // QUERY
    const GET_ALL_USERS_REQUEST = `
      query {
        users {
          firstname
          lastname
          description
          postsWritten {
            id
            title
            description
            solution
            idAuthor
            totalClaps
          }
        }
      }
    `;

    const { data: users } = await axios
      .post(URL_GRAPHQL_API_TEST_INTEGRATION, { query: GET_ALL_USERS_REQUEST });

    expect(users).toMatchSnapshot();
  });
});
