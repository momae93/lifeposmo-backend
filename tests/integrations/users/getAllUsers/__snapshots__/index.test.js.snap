// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[INTEGRATION][USERS] : Get all user should success returns all fields when calling * users * with all its fields 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "birthdate": "1996-09-27",
        "description": "Description of Mickael Au",
        "firstname": "Mickael",
        "id": "1",
        "isMale": true,
        "lastname": "AU",
        "password": "NotSoEncryptedPwd",
        "username": "mickaelau",
      },
      Object {
        "birthdate": "1993-03-11",
        "description": "Description of John Doe",
        "firstname": "John",
        "id": "2",
        "isMale": true,
        "lastname": "DOE",
        "password": "NotSoEncryptedPwd",
        "username": "johndoe",
      },
    ],
  },
}
`;

exports[`[INTEGRATION][USERS] : Get all user should success returns all fields when calling * users * with their postwritten 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "description": "Description of Mickael Au",
        "firstname": "Mickael",
        "lastname": "AU",
        "postsWritten": Array [
          Object {
            "description": "This a description",
            "id": "1",
            "idAuthor": 1,
            "solution": "This is a solution",
            "title": "This is the title",
            "totalClaps": 0,
          },
        ],
      },
      Object {
        "description": "Description of John Doe",
        "firstname": "John",
        "lastname": "DOE",
        "postsWritten": Array [],
      },
    ],
  },
}
`;

exports[`[INTEGRATION][USERS] : Get all user should success returns some fields when calling * users * with some of its fields 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "description": "Description of Mickael Au",
        "firstname": "Mickael",
        "lastname": "AU",
      },
      Object {
        "description": "Description of John Doe",
        "firstname": "John",
        "lastname": "DOE",
      },
    ],
  },
}
`;
